flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/FileArchivedExecutionGraphStore.java

```diff
@@ -0,0 +1,298 @@
+		// shutdown hook itself
+		if (shutdownHook != null && shutdownHook != Thread.currentThread()) {
+			try {
+				Runtime.getRuntime().removeShutdownHook(shutdownHook);
+			}
+			catch (IllegalStateException e) {
+				// race, JVM is in shutdown already, we can safely ignore this
+			}
+			catch (Throwable t) {
+				LOG.warn("Exception while unregistering FileArchivedExecutionGraphStore's cleanup shutdown hook.", t);
+			}
+		}
+	}
+
+	// --------------------------------------------------------------
+	// Internal methods
+	// --------------------------------------------------------------
+
+	private int calculateSize(JobID jobId, ArchivedExecutionGraph serializableExecutionGraph) {
+		final File archivedExecutionGraphFile = getExecutionGraphFile(jobId);
+
+		if (archivedExecutionGraphFile.exists()) {
+			return Math.toIntExact(archivedExecutionGraphFile.length());
+		} else {
+			LOG.debug("Could not find archived execution graph file for {}. Estimating the size instead.", jobId);
+			return serializableExecutionGraph.getAllVertices().size() * 1000 +
+				serializableExecutionGraph.getAccumulatorsSerialized().size() * 1000;
+		}
+	}
+
+	private ArchivedExecutionGraph loadExecutionGraph(JobID jobId) throws IOException, ClassNotFoundException {
+		final File archivedExecutionGraphFile = getExecutionGraphFile(jobId);
+
+		if (archivedExecutionGraphFile.exists()) {
+			try (FileInputStream fileInputStream = new FileInputStream(archivedExecutionGraphFile)) {
+				return InstantiationUtil.deserializeObject(fileInputStream, getClass().getClassLoader());
+			}
+		} else {
+			throw new FileNotFoundException("Could not find file for archived execution graph " + jobId +
+				". This indicates that the file either has been deleted or never written.");
+		}
+	}
+
+	private void storeArchivedExecutionGraph(ArchivedExecutionGraph archivedExecutionGraph) throws IOException {
+		final File archivedExecutionGraphFile = getExecutionGraphFile(archivedExecutionGraph.getJobID());
+
+		try (FileOutputStream fileOutputStream = new FileOutputStream(archivedExecutionGraphFile)) {
+			InstantiationUtil.serializeObject(fileOutputStream, archivedExecutionGraph);
+		}
+	}
+
+	private File getExecutionGraphFile(JobID jobId) {
+		return new File(storageDir, jobId.toString());
+	}
+
+	private void deleteExecutionGraphFile(JobID jobId) {
+		Preconditions.checkNotNull(jobId);
+
+		final File archivedExecutionGraphFile = getExecutionGraphFile(jobId);
+
+		try {
+			FileUtils.deleteFileOrDirectory(archivedExecutionGraphFile);
+		} catch (IOException e) {
+			LOG.debug("Could not delete file {}.", archivedExecutionGraphFile, e);
+		}
+
+		archivedExecutionGraphCache.invalidate(jobId);
+		jobDetailsCache.invalidate(jobId);
+	}
+
+	private static File initExecutionGraphStorageDirectory(File tmpDir) throws IOException {
+		final int maxAttempts = 10;
+
+		for (int attempt = 0; attempt < maxAttempts; attempt++) {
```

Comment:
In case of collisions this will decrease a bit the probability, even though it won't be `0`. This is similar to how we do it for the `BlobServer` but strictly speaking most likely not necessary.

Reply:
The UUID is generated using a cryptographically strong random number generator. There should be enough entropy to ensure no collisions will happen. But I am also okay with leaving it to keep it consistent with BlobServer.